server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -744,6 +757,32 @@ public DataSegment map(int index, ResultSet r, StatementContext ctx) throws SQLE
 
     // Replace "dataSources" atomically.
     dataSources = newDataSources;
+    overshadowedSegments = ImmutableSet.copyOf(determineOvershadowedSegments(segments));
```

Comment:
> when some dataSources are enabled or disabled outside doPoll

Yes. Also note that there are additional changes here: #7653 (see section "SQLMetadataSegmentManager: remove data from dataSources in markAsUnusedSegmentsInInterval and markSegmentsAsUnused"), which were missed in #7490.

> Also, if they do become invalid, would a comment be enough, or something should be done in code to prevent invalid overshadowed segments ?

I think `overshadowedSegments` should become invalidated.

Reply:
btw, i reverted the above mentioned change and added `DataSourcesSnapshot`